profiling/profile_hooks.py

```diff
@@ -0,0 +1,230 @@
+    parser.add_argument("--dtypes", type=str, help="Comma-separated dtypes")
+    parser.add_argument("--model_dim", type=int)
+    parser.add_argument("--log_level", type=str, default="warning")
+    parser.add_argument("--single_gpu_only", action="store_true")
+    parser.add_argument("--multi_gpu_only", action="store_true")
+    parser.add_argument("--n_layers", type=int, default=32)
+    parser.add_argument("--debug", action="store_true")
+
+    parser = _add_profile_args(parser)
+    parser = _add_distributed_args(parser)
+
+    args = parser.parse_args()
+
+    args = validate_args(args)
+
+    return args
+
+
+def validate_args(args):
+    # Manual override for one dtype.
+    if args.dtypes is not None and args.dtypes in DTYPES:
+        dtypes = []
+        for d in args.dtypes.split(","):
+            if d in DTYPES:
+                dtypes.append(DTYPES[d])
+            else:
+                raise Exception(f"Unsupported dtype provided: {d}")
+        args.dtype_sweep = dtypes
+    else:
+        args.dtype_sweep = list(DTYPES.values())
+
+    # Manual override for one model dim.
+    if args.model_dim is not None:
+        args.model_dim_sweep = [args.model_dim]
+    else:
+        args.model_dim_sweep = [2 ** i for i in range(9, 15)]
+
+    # Debug overrides both dtype and model dim to test all configurations.
+    if args.debug:
+        args.dtype_sweep = [v for v in DTYPES.values()]
+        args.model_dim_sweep = [16, 32]
+
+    # Determine which experiments to run.
+    assert not (
+        args.single_gpu_only and args.multi_gpu_only
+    ), f"Cannot have both single gpu and multi gpu only flags both True."
+
+    # TP is across all gpus by default.
+    if args.tp_size is None:
+        args.tp_size = int(os.environ["WORLD_SIZE"])
+
+    # Make folder for profiling.
+    if args.profile_save_profile:
+        if args.profile_dir is None:
+            args.profile_dir = (
+                f"{os.getcwd()}/profiles/profile_{os.environ['SLURM_JOB_ID']}"
+            )
+        if not os.path.isdir(args.profile_dir):
+            os.makedirs(args.profile_dir, exist_ok=True)
+
+    return args
+
+
+def print_args(args):
+    pp = pprint.PrettyPrinter(width=80)
+    pp.pprint(vars(args))
+
+
+def main(args):
+    setup_logger(args.log_level)
+
+    # Silence kineto warnings.
+    os.environ["KINETO_LOG_LEVEL"] = "5"
+
+    # Initialize distributed and megatron-lm parallel state.
+    init_megatron_dist(args)
+    rank = torch.distributed.get_rank()
+    torch.manual_seed(rank)
+    if rank == 0:
+        print_args(args)
+
+    # Construct experiment setup functions and create profile folders.
+    manager = ExperimentNetworkManager()
+    experiments = manager.named_experiments
+
+    prefix = ""
+    if args.single_gpu_only:
+        prefix = "single_gpu"
+    elif args.multi_gpu_only:
+        prefix = "multi_gpu"
+    experiments = [getattr(manager, e) for e in experiments if e.startswith(prefix)]
```

Comment:
This can probably be pushed inside `ExperimentNetworkManager`, where we just request experiment handles via prefix-matching.